File: app/src/main/java/com/mavenkalabs/adskipper/rules/RulesParser.java
package com.mavenkalabs.adskipper.rules;

import java.util.Arrays;

public class RulesParser {
    public BaseRule parse(String ruleAsString, String packageName) {
        if (ruleAsString == null || ruleAsString.trim().isEmpty()) {
            throw new IllegalArgumentException();
        } else if (ruleAsString.contains("&")) {
            return new MultiRule(MultiRule.Conditionality.AND,
                    Arrays.stream(ruleAsString.split("&"))
                            .map((s) -> parse(s, packageName))
                            .toArray(BaseRule[]::new));
        } else {
            if (ruleAsString.startsWith(RuleConstants.RULE_ID_NO_RECENT_USER_CLICK)) {
                return new NoRecentUserClickRule(ruleAsString);
            } else if (ruleAsString.startsWith("!")) {
                return new MustNotExistRule(ruleAsString.substring(1), packageName);
            } else {
                return new MustExistRule(ruleAsString, packageName);
            }
        }
    }

    public BaseRule orRules(BaseRule... rules) {
        if (rules == null || rules.length == 0) {
            throw new IllegalArgumentException();
        } else if (rules.length == 1) {
            return rules[0];
        } else {
            return new MultiRule(MultiRule.Conditionality.OR, rules);
        }
    }
}

File: app/src/main/kotlin/com/mavenkalabs/adskipper/rules/RulesParser.kt
package com.mavenkalabs.adskipper.rules

class RulesParser {
    fun parse(ruleAsString: String, packageName: String): BaseRule {
        require(ruleAsString.trim().isNotEmpty())
        return if (ruleAsString.contains("&")) {
            MultiRule (
                MultiRule.Conditionality.AND,
                *ruleAsString.split("&".toRegex())
                    .map { s -> parse(s, packageName)  }
                    .toTypedArray())
        } else {
            if (ruleAsString.startsWith(RULE_ID_NO_RECENT_USER_CLICK)) {
                NoRecentUserClickRule(ruleAsString)
            } else if (ruleAsString.startsWith("!")) {
                MustNotExistRule(ruleAsString.substring(1), packageName)
            } else {
                MustExistRule(ruleAsString, packageName)
            }
        }
    }

    fun orRules(vararg rules: BaseRule): BaseRule {
        require(rules.isNotEmpty())
        return if (rules.size == 1) {
            rules.first()
        } else {
            MultiRule(MultiRule.Conditionality.OR, *rules)
        }
    }
}
